src/services/services.ts

```diff
@@ -2478,6 +2481,57 @@ export function createLanguageService(
         }
     }
 
+    function getLinkedEditingRangeAtPosition(fileName: string, position: number): LinkedEditingInfo | undefined {
+        const sourceFile = syntaxTreeCache.getCurrentSourceFile(fileName);
+        const token = findPrecedingToken(position, sourceFile);
+        if (!token || token.parent.kind === SyntaxKind.SourceFile) return undefined;
+
+        if (isJsxFragment(token.parent.parent)) {
+            const openFragment = token.parent.parent.openingFragment;
+            const closeFragment = token.parent.parent.closingFragment;
+            if (containsParseError(openFragment) || containsParseError(closeFragment)) return undefined;
+
+            const openPos = openFragment.getStart(sourceFile) + 1; // "<".length
+            const closePos = closeFragment.getStart(sourceFile) + 2; // "</".length
+
+            // only allows linked editing right after opening bracket: <| ></| >
+            if ((position !== openPos) && (position !== closePos)) return undefined;
+
+            return { ranges: [{ start: openPos, length: 0 }, { start: closePos, length: 0 }] };
+        }
+        else {
+            // determines if the cursor is in an element tag
+            const element = findAncestor(token.parent,
+                n => {
+                    if (!n.parent) return "quit";
+                    else if (isJsxElement(n.parent)) {
+                        if (isJsxOpeningElement(n) || isJsxClosingElement(n)) {
+                            return true;
+                        }
+                        return "quit";
+                    }
+                    return false;
+                });
+            if (!element || !(isJsxOpeningElement(element) || isJsxClosingElement(element))) return undefined;

```

Comment:
Is `(isJsxOpeningElement(element) || isJsxClosingElement(element))` possibly `false` given what the `findAncestor` call does? Fine to leave in just in case, but I'm curious if this/how this happens

Reply:
Seems to now be mostly for the typeguard feature, as it seems other code I've added since covers these cases. Should I change it to an `as JsxOpeningElement | JsxClosingElement | undefined` assertion to the result of `findAncestor`?